GVFS/GVFS.Common/Cleanup/GitCleanupQueue.cs

```diff
@@ -0,0 +1,155 @@
+﻿using GVFS.Common.Tracing;
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Threading;
+
+namespace GVFS.Common.Cleanup
+{
+    public class GitCleanupQueue
+    {
+        private readonly object threadLock = new object();
+        private readonly object currentStepLock = new object();
+        private GVFSContext context;
+        private ConcurrentQueue<GitCleanupStep> queue;
+        private Thread thread;
+        private GitCleanupStep currentStep;
+        private bool stopping;
+
+        public GitCleanupQueue(GVFSContext context)
+        {
+            this.context = context;
+            this.queue = new ConcurrentQueue<GitCleanupStep>();
+        }
+
+        public void Enqueue(GitCleanupStep step)
+        {
+            this.queue.Enqueue(step);
+
+            lock (this.threadLock)
+            {
+                if (this.thread == null)
+                {
+                    this.thread = new Thread(() => this.RunQueue());
+                    this.thread.IsBackground = true;
+
+                    try
+                    {
+                        this.thread.Start();
+                    }
+                    catch (ThreadStateException e)
+                    {
+                        this.LogError(nameof(GitCleanupQueue), nameof(this.Enqueue), e);
+                    }
+                    catch (OutOfMemoryException e)
+                    {
+                        this.LogError(nameof(GitCleanupQueue), nameof(this.Enqueue), e);
+                    }
+                }

```

Comment:
@derrickstolee I think we can greatly simplify this class by using a BlockingQueue.  I put a PR into your fork showing what i mean [here](https://github.com/derrickstolee/VFSForGit/pull/2/files).

Reply:
Thanks for this! It is a much cleaner approach that I would never create on my own.